File: backend/app/schemas/signal.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class SignalAction(str, Enum):
    buy = "BUY"
    sell = "SELL"
    hold = "HOLD"


class RiskLevel(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class TechnicalIndicators(BaseModel):
    rsi: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    ma50: float | None = None
    ma200: float | None = None


class SignalResponse(BaseModel):
    id: str | None = None
    symbol: str
    signal: SignalAction
    confidence: float = Field(ge=0, le=100)
    reasoning: str
    risk_level: RiskLevel
    indicators: TechnicalIndicators | None = None
    created_at: datetime | None = None
    source: str = "ai"


class SignalHistoryResponse(BaseModel):
    symbol: str
    signals: list[SignalResponse]

File: frontend/src/types/signal.ts
export type SignalAction = "BUY" | "SELL" | "HOLD";
export type RiskLevel = "low" | "medium" | "high";

export interface TechnicalIndicators {
  rsi?: number | null;
  macd?: number | null;
  macd_signal?: number | null;
  ma50?: number | null;
  ma200?: number | null;
}

export interface TradingSignal {
  id?: string | null;
  symbol: string;
  signal: SignalAction;
  confidence: number;
  reasoning: string;
  risk_level: RiskLevel;
  indicators?: TechnicalIndicators | null;
  created_at?: string | null;
  source?: string;
}

export interface SignalHistory {
  symbol: string;
  signals: TradingSignal[];
}
